File: ios/runanywhere-sdks-latest/sdk/runanywhere-flutter/packages/runanywhere/lib/features/voice_agent/services/voice_agent_mic_driver.dart
// SPDX-License-Identifier: Apache-2.0
//
// voice_agent_mic_driver.dart — audio ingress for the Flutter voice agent.
//
// The C ABI owns NO microphone (rac_voice_agent.h "Audio-Ingress Contract"):
// the platform SDK must capture mic audio and push complete utterances into
// the C core, or the session is dead air. Without this driver the Flutter
// voice agent only subscribed to the output event stream and never fed any
// PCM, so VAD/STT never saw audio → the LLM got no input → no reply.
//
// Mirrors Kotlin `VoiceAgentMicDriver`: capture 16 kHz mono PCM16 via
// [AudioCaptureManager], segment utterances with energy-based endpointing,
// and drive each utterance through `rac_voice_agent_process_turn_proto`
// (DartBridgeVoiceAgent.processTurnStream). The turn's VoiceEvents are
// forwarded to [events] (so the public eventStream observes them) and the
// synthesized TTS reply is played through [AudioPlaybackManager].
//
// The endpointing is intentionally simple — the C++ pipeline re-runs its own
// VAD over each submitted buffer; the only job here is deciding where one
// utterance ends. Mic chunks that arrive while a turn is processing are
// dropped: the pipeline is strictly turn-taking (no barge-in), which also
// avoids transcribing the device's own TTS output.

import 'dart:async';
import 'dart:math';
import 'dart:typed_data';

import 'package:runanywhere/features/stt/services/audio_capture_manager.dart';
import 'package:runanywhere/features/tts/services/audio_playback_manager.dart';
import 'package:runanywhere/foundation/logging/sdk_logger.dart';
import 'package:runanywhere/generated/voice_agent_service.pb.dart'
    as voice_agent_pb;
import 'package:runanywhere/generated/voice_events.pb.dart' as voice_events_pb;
import 'package:runanywhere/native/dart_bridge.dart';
import 'package:runanywhere/native/dart_bridge_audio.dart';

/// Captures mic audio and drives per-utterance voice-agent turns. [start]
/// begins capture; [events] streams every turn's VoiceEvents; [stop] tears
/// capture + playback down. Mirrors Kotlin `VoiceAgentMicDriver`.
class VoiceAgentMicDriver {
  VoiceAgentMicDriver();

  static final _logger = SDKLogger('VoiceAgentMic');

  static const int _sampleRateHz = 16000;
  static const int _bytesPerSample = 2;

  /// Absolute floor for the adaptive speech threshold (normalized RMS).
  static const double _speechRmsThreshold = 0.015;

  /// Speech must exceed this multiple of the tracked ambient noise floor.
  static const double _speechFloorMultiplier = 2.2;

  /// Per-chunk rate at which the ambient floor creeps up toward louder ambient.
  static const double _noiseFloorRise = 0.05;

  /// Trailing silence that closes an utterance.
  static const int _endOfUtteranceSilenceMs = 800;

  /// Utterances with less accumulated speech than this are noise.
  static const int _minSpeechMs = 300;

  /// Hard cap so a noisy room cannot grow an unbounded buffer.
  static const int _maxUtteranceMs = 15000;

  /// Leading chunks kept so the utterance onset is not clipped.
  static const int _preRollChunks = 3;

  /// Piper's native rate; used when an audio frame omits sample_rate_hz.
  static const int _defaultTtsSampleRateHz = 22050;

  final AudioCaptureManager _capture = AudioCaptureManager();
  final AudioPlaybackManager _playback = AudioPlaybackManager();
  final StreamController<voice_events_pb.VoiceEvent> _out =
      StreamController<voice_events_pb.VoiceEvent>();

  StreamSubscription<Uint8List>? _micSub;
  bool _stopped = false;
  bool _processing = false;

  // Segmentation state.
  final List<Uint8List> _preRoll = <Uint8List>[];
  final BytesBuilder _utterance = BytesBuilder();
  bool _inSpeech = false;
  int _speechMs = 0;
  int _silenceMs = 0;
  double _noiseFloor = _speechRmsThreshold;

  /// VoiceEvents produced by each turn (userSaid, llm tokens, audio, pipeline
  /// state). The public eventStream yields from this.
  Stream<voice_events_pb.VoiceEvent> get events => _out.stream;

  /// Begin mic capture. On permission/capture failure the [events] stream is
  /// closed with an error so the collector can surface it.
  Future<void> start() async {
    // The voice agent runs a single full-duplex (.playAndRecord) session for the
    // whole turn-taking loop — the `record` plugin configures it on
    // startRecording (defaultToSpeaker by default). Playback must NOT switch the
    // session to the output-only `.playback` category, or it trips
    // AVAudioSessionErrorInsufficientPriority ('!pri', OSStatus 561017449) and
    // the reply is dropped. Mirrors the iOS Swift driver
    // (playback.managesAudioSession = false).
    _playback.managesAudioSession = false;
    final stream =
        await _capture.startRecording(sampleRate: _sampleRateHz, numChannels: 1);
    if (stream == null) {
      _out.addError(StateError(
          'Microphone capture unavailable (permission denied or busy)'));
      unawaited(_out.close());
      return;
    }
    _logger.info('Voice-agent mic capture started');
    _micSub = stream.listen(
      _onChunk,
      onError: (Object e, StackTrace st) => _logger.warning('Mic error: $e'),
      cancelOnError: false,
    );
  }

  /// Stop capture + playback and close the event stream.
  Future<void> stop() async {
    if (_stopped) return;
    _stopped = true;
    await _micSub?.cancel();
    _micSub = null;
    await _capture.stopRecording();
    await _playback.stop();
    if (!_out.isClosed) {
      await _out.close();
    }
    _logger.info('Voice-agent mic capture stopped');
  }

  void _onChunk(Uint8List chunk) {
    // Drop chunks while a turn is processing (turn-taking, no barge-in).
    if (_stopped || _processing || chunk.isEmpty) return;

    final chunkMs = chunk.length * 1000 ~/ (_sampleRateHz * _bytesPerSample);
    final level = _rms(chunk);
    // Adaptive endpointing: track the ambient floor — drop instantly to any
    // quieter level, creep up only while not in speech — and require a chunk
    // to rise clearly above that floor to count as speech.
    final speechThreshold =
        max(_speechRmsThreshold, _noiseFloor * _speechFloorMultiplier);
    final isSpeech = level >= speechThreshold;
    if (level < _noiseFloor) {
      _noiseFloor = level;
    } else if (!isSpeech) {
      _noiseFloor += (level - _noiseFloor) * _noiseFloorRise;
    }

    if (!_inSpeech) {
      _preRoll.add(chunk);
      while (_preRoll.length > _preRollChunks) {
        _preRoll.removeAt(0);
      }
      if (isSpeech) {
        _inSpeech = true;
        _speechMs = chunkMs;
        _silenceMs = 0;
        _utterance.clear();
        for (final pre in _preRoll) {
          _utterance.add(pre);
        }
        _preRoll.clear();
      }
      return;
    }

    _utterance.add(chunk);
    if (isSpeech) {
      _speechMs += chunkMs;
      _silenceMs = 0;
    } else {
      _silenceMs += chunkMs;
    }

    final utteranceMs =
        _utterance.length * 1000 ~/ (_sampleRateHz * _bytesPerSample);
    if (_silenceMs >= _endOfUtteranceSilenceMs || utteranceMs >= _maxUtteranceMs) {
      final audio = _utterance.toBytes();
      final hadSpeech = _speechMs >= _minSpeechMs;
      _resetSegmentation();
      if (hadSpeech) {
        unawaited(_processTurn(audio));
      } else {
        _logger.debug('Utterance discarded (${_speechMs}ms speech < '
            '${_minSpeechMs}ms)');
      }
    }
  }

  void _resetSegmentation() {
    _inSpeech = false;
    _speechMs = 0;
    _silenceMs = 0;
    _utterance.clear();
    _preRoll.clear();
  }

  Future<void> _processTurn(Uint8List audio) async {
    _processing = true;
    final ttsPcm = BytesBuilder();
    var ttsSampleRate = 0;
    var ttsEncoding = voice_events_pb.AudioEncoding.AUDIO_ENCODING_UNSPECIFIED;
    try {
      final request = voice_agent_pb.VoiceAgentTurnRequest(
        requestId: 'turn-${DateTime.now().microsecondsSinceEpoch}',
        audioData: audio,
        sampleRateHz: _sampleRateHz,
        channels: 1,
        encoding: voice_events_pb.AudioEncoding.AUDIO_ENCODING_PCM_S16_LE,
      );
      _logger.info('Submitting voice turn (${audio.length} bytes)');

      await for (final ev in DartBridge.voiceAgent.processTurnStream(request)) {
        if (!_out.isClosed) {
          _out.add(ev);
        }
        if (ev.hasAudio() && ev.audio.pcm.isNotEmpty) {
          ttsPcm.add(ev.audio.pcm is Uint8List
              ? ev.audio.pcm as Uint8List
              : Uint8List.fromList(ev.audio.pcm));
          if (ev.audio.sampleRateHz > 0) {
            ttsSampleRate = ev.audio.sampleRateHz;
          }
          if (ev.audio.encoding !=
              voice_events_pb.AudioEncoding.AUDIO_ENCODING_UNSPECIFIED) {
            ttsEncoding = ev.audio.encoding;
          }
        }
      }
    } catch (e) {
      _logger.warning('Voice turn failed: $e');
      if (!_out.isClosed) {
        _out.addError(e);
      }
    } finally {
      _processing = false;
      _resetSegmentation();
    }

    await _playTts(ttsPcm.toBytes(), ttsSampleRate, ttsEncoding);
  }

  // Play the turn's synthesized reply through the TTS sink. Runs before the
  // next mic chunk is processed (the _processing gate is cleared above but the
  // playback await keeps the turn logically open), so the mic stays gated
  // while the device speaks — no self-transcription.
  Future<void> _playTts(
    Uint8List pcm,
    int sampleRateHz,
    voice_events_pb.AudioEncoding encoding,
  ) async {
    if (pcm.isEmpty || _stopped) return;
    final sampleRate = sampleRateHz > 0 ? sampleRateHz : _defaultTtsSampleRateHz;
    // TTS backends emit f32 LE by default (AudioFrameEvent contract); only
    // convert as PCM16 when the frame explicitly says so.
    final wav =
        encoding == voice_events_pb.AudioEncoding.AUDIO_ENCODING_PCM_S16_LE
            ? DartBridgeAudio.int16ToWav(pcm, sampleRate)
            : DartBridgeAudio.float32ToWav(pcm, sampleRate);
    if (wav == null || wav.isEmpty) {
      _logger.warning('TTS audio conversion failed (${pcm.length} bytes, '
          '${sampleRate}Hz, $encoding)');
      return;
    }
    _processing = true; // keep mic gated while speaking
    try {
      await _playback.play(wav);
    } catch (e) {
      _logger.warning('Agent reply playback failed: $e');
    } finally {
      _processing = false;
      _resetSegmentation();
    }
  }

  double _rms(Uint8List chunk) {
    final samples = chunk.length ~/ _bytesPerSample;
    if (samples == 0) return 0.0;
    final data = ByteData.sublistView(chunk);
    var sum = 0.0;
    for (var i = 0; i < samples; i++) {
      final sample = data.getInt16(i * _bytesPerSample, Endian.little).toDouble();
      sum += sample * sample;
    }
    return sqrt(sum / samples) / 32767.0;
  }
}

File: ios/runanywhere-sdks-latest/sdk/runanywhere-react-native/packages/core/src/Features/VoiceSession/VoiceAgentMicDriver.ts
/**
 * VoiceAgentMicDriver.ts — microphone audio ingress for the RN voice agent.
 *
 * The commons C ABI owns NO microphone (rac_voice_agent.h "audio-ingress
 * contract"): subscribing to `streamVoiceAgent()` alone is dead air. The
 * platform SDK must capture mic audio and push utterances into the core, or
 * the pipeline never sees PCM → VAD/STT get nothing → the LLM has no input →
 * no reply. This driver closes that gap for RN.
 *
 * Mirrors the Kotlin/Flutter `VoiceAgentMicDriver`: capture 16 kHz mono PCM16
 * via {@link AudioCaptureManager}, segment utterances with energy-based
 * endpointing, and run each utterance through `processVoiceTurn` (the one-shot
 * VAD→STT→LLM→TTS C entry, which on this build also re-runs the loaded VAD
 * model over the submitted buffer). The transcription + assistant response are
 * surfaced via {@link VoiceAgentMicCallbacks.onTurn}, and the synthesized TTS
 * reply is played through {@link AudioPlaybackManager}.
 *
 * The endpointing is intentionally coarse — it only decides where one
 * utterance ends; the C++ pipeline VADs each submitted buffer. Mic chunks that
 * arrive while a turn is processing (or while the reply is playing) are
 * dropped: the pipeline is strictly turn-taking (no barge-in), which also
 * avoids transcribing the device's own TTS output.
 */

import { AudioCaptureManager } from './AudioCaptureManager';
import { AudioPlaybackManager } from './AudioPlaybackManager';
import { processVoiceTurn } from '../../Public/Extensions/VoiceAgent/RunAnywhere+VoiceAgent';
import { SDKLogger } from '../../Foundation/Logging/Logger/SDKLogger';

const logger = new SDKLogger('VoiceAgentMicDriver');

/** A completed turn: what the user said and how the agent replied. */
export interface VoiceAgentMicTurn {
  userText: string;
  assistantText: string;
}

/** Coarse pipeline phase, for UI status. */
export type VoiceAgentMicPhase = 'listening' | 'processing' | 'speaking';

export interface VoiceAgentMicCallbacks {
  /** A turn finished: user transcription + assistant reply. */
  onTurn?: (turn: VoiceAgentMicTurn) => void;
  /** Coarse phase transitions (listening / processing / speaking). */
  onPhase?: (phase: VoiceAgentMicPhase) => void;
  /** A non-fatal turn error (capture continues). */
  onError?: (error: Error) => void;
}

const SAMPLE_RATE_HZ = 16000;
const BYTES_PER_SAMPLE = 2;

/** Absolute floor for the adaptive speech threshold (normalized RMS). */
const SPEECH_RMS_THRESHOLD = 0.015;
/** Speech must exceed this multiple of the tracked ambient noise floor. */
const SPEECH_FLOOR_MULTIPLIER = 2.2;
/** Per-chunk rate at which the ambient floor creeps toward louder ambient. */
const NOISE_FLOOR_RISE = 0.05;
/** Trailing silence that closes an utterance. */
const END_OF_UTTERANCE_SILENCE_MS = 800;
/** Utterances with less accumulated speech than this are discarded as noise. */
const MIN_SPEECH_MS = 300;
/** Hard cap so a noisy room cannot grow an unbounded buffer. */
const MAX_UTTERANCE_MS = 15000;
/** Leading chunks kept so the utterance onset is not clipped. */
const PRE_ROLL_CHUNKS = 3;

export class VoiceAgentMicDriver {
  private readonly capture = new AudioCaptureManager();
  private readonly playback = new AudioPlaybackManager();
  private callbacks: VoiceAgentMicCallbacks = {};

  private stopped = false;
  private processing = false;

  // Segmentation state.
  private preRoll: Uint8Array[] = [];
  private utterance: Uint8Array[] = [];
  private utteranceBytes = 0;
  private inSpeech = false;
  private speechMs = 0;
  private silenceMs = 0;
  private noiseFloor = SPEECH_RMS_THRESHOLD;

  /**
   * Request mic permission and begin capture. Returns `false` if the
   * permission was denied (caller should surface it); throws on capture
   * failure.
   */
  async start(callbacks: VoiceAgentMicCallbacks = {}): Promise<boolean> {
    this.callbacks = callbacks;
    this.stopped = false;

    const granted = await this.capture.requestPermission();
    if (!granted) {
      return false;
    }

    // The voice agent runs a single full-duplex (.playAndRecord) session for the
    // whole turn-taking loop so TTS replies can play through the speaker while
    // the mic session stays live. Activate it BEFORE startRecording so capture
    // reuses it instead of forcing the output-only .record session — under
    // .record, playback cannot claim the output and trips
    // AVAudioSessionErrorInsufficientPriority ('!pri', OSStatus 561017449).
    // Mirrors the iOS Swift driver's configureVoiceAudioSession(); no-op on
    // Android.
    await this.capture.activateAudioSession();
    await this.capture.startRecording((chunk) => this.onChunk(chunk));
    this.callbacks.onPhase?.('listening');
    logger.info('Voice-agent mic capture started');
    return true;
  }

  /** Stop capture + playback and reset segmentation. */
  stop(): void {
    if (this.stopped) return;
    this.stopped = true;
    try {
      this.capture.stopRecording();
    } catch (e) {
      logger.warning(`stopRecording failed: ${String(e)}`);
    }
    this.playback.stop();
    this.resetSegmentation();
    logger.info('Voice-agent mic capture stopped');
  }

  private onChunk(chunk: Uint8Array): void {
    // Drop chunks while a turn is processing or the reply is playing
    // (turn-taking, no barge-in, no self-transcription).
    if (this.stopped || this.processing || chunk.length === 0) return;

    const chunkMs = (chunk.length * 1000) / (SAMPLE_RATE_HZ * BYTES_PER_SAMPLE);
    const level = this.rms(chunk);

    // Adaptive endpointing: track the ambient floor (drop instantly to any
    // quieter level, creep up only while not in speech) and require a chunk to
    // rise clearly above that floor to count as speech.
    const speechThreshold = Math.max(
      SPEECH_RMS_THRESHOLD,
      this.noiseFloor * SPEECH_FLOOR_MULTIPLIER
    );
    const isSpeech = level >= speechThreshold;
    if (level < this.noiseFloor) {
      this.noiseFloor = level;
    } else if (!isSpeech) {
      this.noiseFloor += (level - this.noiseFloor) * NOISE_FLOOR_RISE;
    }

    if (!this.inSpeech) {
      this.preRoll.push(chunk);
      while (this.preRoll.length > PRE_ROLL_CHUNKS) {
        this.preRoll.shift();
      }
      if (isSpeech) {
        this.inSpeech = true;
        this.speechMs = chunkMs;
        this.silenceMs = 0;
        this.utterance = [...this.preRoll];
        this.utteranceBytes = this.preRoll.reduce((n, c) => n + c.length, 0);
        this.preRoll = [];
      }
      return;
    }

    this.utterance.push(chunk);
    this.utteranceBytes += chunk.length;
    if (isSpeech) {
      this.speechMs += chunkMs;
      this.silenceMs = 0;
    } else {
      this.silenceMs += chunkMs;
    }

    const utteranceMs =
      (this.utteranceBytes * 1000) / (SAMPLE_RATE_HZ * BYTES_PER_SAMPLE);
    if (
      this.silenceMs >= END_OF_UTTERANCE_SILENCE_MS ||
      utteranceMs >= MAX_UTTERANCE_MS
    ) {
      const audio = this.concatUtterance();
      const hadSpeech = this.speechMs >= MIN_SPEECH_MS;
      this.resetSegmentation();
      if (hadSpeech) {
        void this.processTurn(audio);
      } else {
        logger.debug(
          `Utterance discarded (${this.speechMs}ms speech < ${MIN_SPEECH_MS}ms)`
        );
      }
    }
  }

  private concatUtterance(): Uint8Array {
    const out = new Uint8Array(this.utteranceBytes);
    let offset = 0;
    for (const c of this.utterance) {
      out.set(c, offset);
      offset += c.length;
    }
    return out;
  }

  private resetSegmentation(): void {
    this.inSpeech = false;
    this.speechMs = 0;
    this.silenceMs = 0;
    this.utterance = [];
    this.utteranceBytes = 0;
    this.preRoll = [];
  }

  private async processTurn(audio: Uint8Array): Promise<void> {
    this.processing = true;
    this.callbacks.onPhase?.('processing');
    logger.info(`Submitting voice turn (${audio.length} bytes)`);
    try {
      const result = await processVoiceTurn(audio);

      const userText = (result.transcription ?? '').trim();
      const assistantText = (result.assistantResponse ?? '').trim();
      if (userText.length > 0 || assistantText.length > 0) {
        this.callbacks.onTurn?.({ userText, assistantText });
      }

      const wav = result.synthesizedAudio;
      if (wav && wav.byteLength > 0 && !this.stopped) {
        this.callbacks.onPhase?.('speaking');
        // `synthesizedAudio` is a complete WAV blob (commons emits WAV, header
        // included), so play it directly — NOT through play(), which would
        // treat the RIFF header + int16 samples as raw float32 PCM and re-wrap
        // it (→ fast/noisy garbage). Copy into a fresh ArrayBuffer-backed view
        // in case the source spans a larger or shared buffer.
        const copy = new Uint8Array(wav.byteLength);
        copy.set(wav);
        try {
          await this.playback.playWav(copy.buffer);
        } catch (e) {
          logger.warning(`Agent reply playback failed: ${String(e)}`);
        }
      }
    } catch (e) {
      const err = e instanceof Error ? e : new Error(String(e));
      logger.warning(`Voice turn failed: ${err.message}`);
      this.callbacks.onError?.(err);
    } finally {
      this.processing = false;
      this.resetSegmentation();
      if (!this.stopped) {
        this.callbacks.onPhase?.('listening');
      }
    }
  }

  private rms(chunk: Uint8Array): number {
    const samples = Math.floor(chunk.length / BYTES_PER_SAMPLE);
    if (samples === 0) return 0;
    const view = new DataView(chunk.buffer, chunk.byteOffset, chunk.byteLength);
    let sum = 0;
    for (let i = 0; i < samples; i++) {
      const sample = view.getInt16(i * BYTES_PER_SAMPLE, true);
      sum += sample * sample;
    }
    return Math.sqrt(sum / samples) / 32767;
  }
}
